File: battle_engine/app/contracts/consumers/mutate_team_contract.rb
# frozen_string_literal: true

module Contracts
  module Consumers
    class MutateTeamContract < Dry::Validation::Contract
      params do
        required(:player_id).filled(:integer)
        required(:name).filled(:string)
        required(:pokemons).array(:hash) do
          required(:pokemon_template_id).filled(:integer)
          optional(:nickname).maybe(:string)
          optional(:gender).maybe(:string)
          optional(:nature).maybe(:string)
          optional(:ivs).maybe(:hash)
          optional(:evs).maybe(:hash)
          required(:moves).array(:integer)
        end
        optional(:timestamp).maybe(:string)
        optional(:team_id).maybe(:integer)
      end
    end
  end
end

File: battle_real_time/lib/battle_real_time/contracts/publishers/player_actions/mutate_team_contract.ex
defmodule BattleRealTime.Contracts.Publishers.PlayerActions.MutateTeamContract do
  @moduledoc """
  Validates the payload for the mutate_team player action, under the player_actions queue.
  """
  use BattleRealTime.Contracts.Contract

  @primary_key false
  embedded_schema do
    field(:player_id, :string)
    field(:name, :string)
    field(:timestamp, :string)
    field(:team_id, :integer)

    embeds_many :pokemons, PokemonItem, primary_key: false do
      field(:pokemon_template_id, :integer)
      field(:nickname, :string)
      field(:nature, :string)
      field(:gender, :string)
      field(:ivs, :map)
      field(:evs, :map)
      field(:moves, {:array, :integer})
    end
  end

  def changeset(struct, params) do
    struct
    |> cast(params, [:player_id, :name, :timestamp, :team_id])
    |> validate_required([:player_id, :name, :pokemons])
    |> cast_embed(:pokemons, required: true, with: &pokemon_changeset/2)
  end

  defp pokemon_changeset(struct, params) do
    struct
    |> cast(params, [:pokemon_template_id, :nickname, :nature, :gender, :ivs, :evs, :moves])
    |> validate_required([:pokemon_template_id, :moves])
  end
end
